src/pie_utils/processor/document/candidate_relation_adder.py

```diff
@@ -0,0 +1,208 @@
+from __future__ import annotations
+
+import json
+import logging
+import random
+from collections import defaultdict
+from typing import Any, Dict, List
+
+from pytorch_ie.annotations import BinaryRelation, Span
+from pytorch_ie.utils.span import is_contained_in
+
+from pie_utils.span.slice import distance
+from pie_utils.statistics import WithStatistics
+
+from ..document import DocumentWithEntitiesRelationsAndPartition
+
+logger = logging.getLogger(__name__)
+
+
+class CandidateRelationAdder(WithStatistics):
+    """CandidateRelationAdder adds binary relations to a document based on various parameters. It
+    goes through combinations of available entity pairs as possible candidates for new relations.
+    Entity pairs which are already part of document as a relation are not added again.
+
+    params:
+        label: label for the new relations to be added
+        use_partition: A boolean parameter to enable partition wise relation creation. If this parameter is enabled then
+                    it uses available partitions in the document otherwise use whole document as single partition
+        max_distance: An optional parameter which restricts the maximum distance between candidate entity pair to form a
+                    new relation. If distance between entity pair is more than the value of this parameter then candidate
+                    pair is discarded.
+        distance_type: type of distance to be calculated between two entities, it can be inner, outer or center. It is
+                    important to note that value for max_distance parameter should be given keeping value of this parameter
+                    in mind.
+        added_rels_upper_bound_factor: It is an optional parameter used to calculate the upper bound for the number of
+                                    new relations that can be added. Upper bound is either total combination possible or
+                                     value of this parameter times number of available relations (whichever is min)
```

Comment:
This is too complicated. short formula would help, I guess (i.e. n = k * ... where n will be the upper bound of added relations). you do not have to say that the min is used (already said by "upper bound").

Reply:
True it was complicated even while writing, will use the short formula. Thanks

Reply:
fixed in [6a5bf0e](https://github.com/ArneBinder/pie-utils/pull/14/commits/6a5bf0e3b0e083b25ff15b1828eb99579208d66f)